timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/HistoryAnalyzer.java

```diff
@@ -36,15 +43,49 @@ private HistoryAnalyzer() {
         // do not create instance of this class
     }
 
+    public static CorruptionHealthReport corruptionHealthReportForHistory(
+            List<CompletePaxosHistoryForNamespaceAndUseCase> history) {
+
+        Map<NamespaceAndUseCase, CorruptionCheckViolation> namespaceAndUseCaseCorruptionCheckViolationMap =
+                history.stream()
+                        .collect(Collectors.toMap(
+                                HistoryAnalyzer::extractNamespaceAndUseCase,
+                                HistoryAnalyzer::violatedCorruptionChecksForNamespaceAndUseCase));
+
+        SetMultimap<CorruptionCheckViolation, NamespaceAndUseCase> entryEntrySetMultimap = KeyedStream.stream(
+                        namespaceAndUseCaseCorruptionCheckViolationMap)
+                .mapEntries((k, v) -> Maps.immutableEntry(v, k))
+                .filterKeys(CorruptionCheckViolation::raiseErrorAlert)
+                .collectToSetMultimap();
+
+        return ImmutableCorruptionHealthReport.builder()
+                .statusesToNamespaceAndUseCase(entryEntrySetMultimap)
+                .build();
+    }
+
+    private static NamespaceAndUseCase extractNamespaceAndUseCase(
+            CompletePaxosHistoryForNamespaceAndUseCase historyForNamespaceAndUseCase) {
+        return ImmutableNamespaceAndUseCase.builder()
+                .namespace(historyForNamespaceAndUseCase.namespace())
+                .useCase(historyForNamespaceAndUseCase.useCase())
+                .build();
+    }
+
     @VisibleForTesting
-    static List<CorruptionCheckViolation> violatedCorruptionChecksForNamespaceAndUseCase(
+    static CorruptionCheckViolation violatedCorruptionChecksForNamespaceAndUseCase(
```

Comment:
I think this should return List or some Collection given the name?

Reply:
Right, the issue is that (and I missed this while writing #5057) these checks are to be executed in an order, e.g. the check for acceptor quorum for a learned value does not hold value if the learners have diverged as we anyway do not have a learner value to compare accepted values against.